base/gmp.jl

```diff
@@ -837,4 +837,112 @@ if Limb === UInt
     end
 end
 
+module MPQ
+
+# Rational{BigFloat}
+import .Base: unsafe_rational, __throw_rational_argerror_zero
+import ..GMP: BigInt, MPZ, Limb
+
+mutable struct _MPQ
+    num_alloc::Cint
+    num_size::Cint
+    num_d::Ptr{Limb}
+    den_alloc::Cint
+    den_size::Cint
+    den_d::Ptr{Limb}
+    # to prevent GC
+    rat::Rational{BigInt}
+end
+
+const mpq_t = Ref{_MPQ}
+
+_MPQ(x::BigInt,y::BigInt) = _MPQ(x.alloc, x.size, x.d,
+                                 y.alloc, y.size, y.d,
+                                 unsafe_rational(BigInt, x, y))
+_MPQ() = _MPQ(BigInt(), BigInt())
+_MPQ(x::Rational{BigInt}) = _MPQ(x.num, x.den)
```

Comment:
Would it make any difference to do:
```jl
_MPQ(x.num.alloc, x.num.size, x.num.d,
            x.den.alloc, x.den.size, x.den.d,
            x)
```
to avoid re-constructing a `Rational{BigInt}`, or is the compiler smart enough to elide that?

Reply:
It shouldn't do, since
```julia
julia> x = big(1//2)
1//2

julia> y = Base.unsafe_rational(BigInt, x.num, x.den)
1//2

julia> x === y
true
```